design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,163 @@
+This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+A user should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation `kubevirt.io/CPUManagerPolicyBetaOptions:full-pcpus-only`:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    metadata:
+      annotations:
+        kubevirt.io/CPUManagerPolicyBetaOptions:full-pcpus-only
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+
+When the VM is configured with `isolateEmulatorThread: true` and the above annotation, the virt-controller will set the virt-launcher pod with an even amount of CPUs, meaning:
+
+| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) |
+|-----------------------------------|------------------|-----------------------|
+| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         |
+
+Notes:
+1. In both scenarios the guest will get the same amount of CPUs it requested.
+2. For simplicity reasons, the housekeeping cgroup will contain only 1 CPU. However, this is open for discussion.
+
+Pros:
+* Cleaner user experience. No prior knowledge of CPUs added behind the scene is needed. 
+* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
+* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs. 
+* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.
+
+Cons:
+* We’re adding yet another semi alpha knob to an already complicated set of knobs.
+
+# Fix Alternatives
+## Alternative #1 Adding `full-pcpus-only` option as a global configuration
+As opposed to the suggested solution where the user needs to add this annotation to every VM, it is possible to make this behavior global. The assumption is that clusters run homogeneous Nodes, so this will eliminate the need to add this annotation every time.
+
+Pros:
+* VM configuration is less complex.
+* Similar to the way default runtimeclass is configured on Kubevirt.
+
+Cons:
+* The suggested annotation is waisting a CPU when used on nodes with hyperthreading disabled. A cluster-wide configuration may be wasteful for heterogeneous clusters, where node-level granularity is needed.
+
+## Alternative #2 Configure both pod and Guest
```

Comment:
@RamLavi This is also doesn't look like an alternative solution, but a suggestion to how to implement the addition of the core so that it will be even.

Reply:
I kinda agree with you here. 
If this alternative is chosen then the Action item is no-op in kubevirt's perspective. I just wanted this option to be heard, as it has its advantages.

Reply:
Thank you for putting it writing, as this is my favorite alternative.